### ISSUE TITLE
skip() from test.pl causes 'fold_constants JMPENV_PUSH' in __WARN__ handler

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#8436]
Migrated from [rt.perl.org#39083](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=39083) (status was 'resolved')

Searchable as RT39083$

**Comment by p5pRT** [comment]
### From david@profane.mongueurs.net

#### Created by david@profane.mongueurs.net


This is a bug report for perl from david@&#8203;profane\.mongueurs\.net\,
generated with the help of perlbug 1\.35 running under perl 5\.9\.4\.


\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-

Given the following program \(a test script that will skip
all remaining tests if a warning is thrown\)&#8203;:

BEGIN \{
&nbsp;    require 'test\.pl';
\}

plan\( tests => 2 \);

BEGIN \{
&nbsp;    $SIG\{\_\_WARN\_\_\} = sub \{
&nbsp;        SKIP&#8203;: \{
&nbsp;            my $nr = 2 \- curr\_test\(\)\+1;
&nbsp;            skip\( "skip remaining $nr tests"\, $nr\);
&nbsp;        \}
&nbsp;        exit\(1\);
&nbsp;    \};
\}

ok\('abc' == 'def'\,'warn'\);
ok\(1\,'dummy'\);
\_\_END\_\_

This will behave  differently according to whether it is run with
\-w on the command line or not \(the "other" perl is 5\.8\.7\)&#8203;:

david@&#8203;profane&#8203;:/home/perl/src/blead/t% perl warnpanic\.t
1\.\.2
ok 1 \- warn
ok 2 \- dummy

david@&#8203;profane&#8203;:/home/perl/src/blead/t% perl \-w warnpanic\.t
ok 1 \# skip&#8203;: skip remaining 2 tests
ok 2 \# skip&#8203;: skip remaining 2 tests

david@&#8203;profane&#8203;:/home/perl/src/blead/t% perl5\.9\.4 warnpanic\.t
1\.\.2
ok 1 \- warn
ok 2 \- dummy

david@&#8203;profane&#8203;:/home/perl/src/blead/t% perl5\.9\.4 \-w warnpanic\.t
ok 1 \# skip&#8203;: skip remaining 2 tests
ok 2 \# skip&#8203;: skip remaining 2 tests
panic&#8203;: fold\_constants JMPENV\_PUSH returned 2\.

If I hoist the C\<require 'test\.pl'> out of the BEGIN\, the
panic also goes away\.

Yes\, I know it is slightly brain\-damaged to be considering
such an approach in the first place\, but I was in fact
rewriting another test \(F\<ext/IPC/SysV/t/ipcsysv\.t> to
use test\.pl and skip nicely if there was a minor problem
of resource acquisition \(no space left on device at test 7\)\.

I wanted to rewrite the test that checks if a SIG\_SYS is
thrown\, and stumbled across this behaviour \(when changing
$SIG\{SYS\} to $SIG\{\_\_WARN\_\_\} in order to test out the
changes\)\.

Thanks\,
David

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

Site configuration information for perl 5.9.4:

Configured by david at Fri May  5 10:44:55 CEST 2006.

Summary of my perl5 (revision 5 version 9 subversion 4 patch 28103) configuration:
  Platform:
    osname=freebsd, osvers=6.0-stable, archname=i386-freebsd-thread-multi-64int
    uname='freebsd profane.mongueurs.net 6.0-stable freebsd 6.0-stable #5: fri feb 24 11:12:56 cet 2006 root@profane.mongueurs.net:usrobjusrsrcsysprofane i386 '
    config_args='-de -Dusedevel -Duse64bitint -Duseithreads -Accflags=-DPERL_DONT_CREATE_GVSV'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=define, usemultiplicity=define
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=define, use64bitall=undef, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-DHAS_FPSETMASK -DHAS_FLOATINGPOINT_H -DPERL_DONT_CREATE_GVSV -fno-strict-aliasing -pipe -Wdeclaration-after-statement -I/usr/local/include',
    optimize='-O',
    cppflags='-DHAS_FPSETMASK -DHAS_FLOATINGPOINT_H -DPERL_DONT_CREATE_GVSV -fno-strict-aliasing -pipe -Wdeclaration-after-statement -I/usr/local/include'
    ccversion='', gccversion='3.4.4 [FreeBSD] 20050518', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=12345678
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='cc', ldflags ='-pthread -Wl,-E  -L/usr/local/lib'
    libpth=/usr/lib /usr/local/lib
    libs=-lm -lcrypt
    perllibs=-lm -lcrypt
    libc=, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version=''
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags=' '
    cccdlflags='-DPIC -fPIC', lddlflags='-shared  -L/usr/local/lib'

Locally applied patches:
    


@INC for perl 5.9.4:
    /usr/local/lib/perl5/5.9.4/i386-freebsd-thread-multi-64int
    /usr/local/lib/perl5/5.9.4
    /usr/local/lib/perl5/site_perl/5.9.4/i386-freebsd-thread-multi-64int
    /usr/local/lib/perl5/site_perl/5.9.4
    /usr/local/lib/perl5/site_perl
    .


Environment for perl 5.9.4:
    HOME=/home/david
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/sbin:/bin:/usr/sbin:/usr/bin:/usr/games:/usr/local/sbin:/usr/local/bin:/usr/X11R6/bin:/home/david/bin:.
    PERL_BADLANG (unset)
    SHELL=/usr/local/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @rgs

david wrote&#8203;:
> 
> Given the following program \(a test script that will skip
> all remaining tests if a warning is thrown\)&#8203;:

Side note\, that's why the "Bail out\!" magic\-message feature was
implemented in T&#8203;::H and in t/TEST\.

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @nwc10

On Fri\, May 05\, 2006 at 04&#8203;:24&#8203;:17AM \-0700\, david @&#8203; profane\. mongueurs\. net wrote&#8203;:

> panic&#8203;: fold\_constants JMPENV\_PUSH returned 2\.

Mmm\, yes\. I think the minimal test case is&#8203;:

$ \./perl \-lwe 'BEGIN \{$SIG\{\_\_WARN\_\_\} = sub\{exit\}\} "a" == "b"'
panic&#8203;: fold\_constants JMPENV\_PUSH returned 2\.

Presumably the constant folding code should ignore exit\, as it's only an
optimisation? ie this could should print "Here\!"&#8203;:

perl \-lwe 'BEGIN \{$SIG\{\_\_WARN\_\_\} = sub\{exit\}\} if \(0\) \{"a" == "b"\}; print "Here\!"'

Nicholas Clark

**Comment by p5pRT** [comment]
### From david@landgren.net

Rafael Garcia\-Suarez wrote&#8203;:
> david wrote&#8203;:
>> Given the following program \(a test script that will skip
>> all remaining tests if a warning is thrown\)&#8203;:
> 
> Side note\, that's why the "Bail out\!" magic\-message feature was
> implemented in T&#8203;::H and in t/TEST\.

/me reads the source\.

Oh\, well that makes things a lot easier\!

Thanks\,
DAvid
\-\- 
"It's overkill of course\, but you can never have too much overkill\."

**Comment by p5pRT** [comment]
### From @ysth

On Fri\, May 05\, 2006 at 01&#8203;:43&#8203;:37PM \+0100\, Nicholas Clark wrote&#8203;:
> On Fri\, May 05\, 2006 at 04&#8203;:24&#8203;:17AM \-0700\, david @&#8203; profane\. mongueurs\. net wrote&#8203;:
> 
> > panic&#8203;: fold\_constants JMPENV\_PUSH returned 2\.
> 
> Mmm\, yes\. I think the minimal test case is&#8203;:
> 
> $ \./perl \-lwe 'BEGIN \{$SIG\{\_\_WARN\_\_\} = sub\{exit\}\} "a" == "b"'
> panic&#8203;: fold\_constants JMPENV\_PUSH returned 2\.
> 
> Presumably the constant folding code should ignore exit\, as it's only an
> optimisation? ie this could should print "Here\!"&#8203;:
> 
> perl \-lwe 'BEGIN \{$SIG\{\_\_WARN\_\_\} = sub\{exit\}\} if \(0\) \{"a" == "b"\}; print "Here\!"'

We've talked before about bailing out of constant\-folding on a croak;
maybe do the same for any warning?  \(Harder to do\, though\.\)

**Comment by p5pRT** [comment]
### From @iabyn

On Sat\, May 06\, 2006 at 11&#8203;:29&#8203;:16PM \-0700\, Yitzchak Scott\-Thoennes wrote&#8203;:
> On Fri\, May 05\, 2006 at 01&#8203;:43&#8203;:37PM \+0100\, Nicholas Clark wrote&#8203;:
> > On Fri\, May 05\, 2006 at 04&#8203;:24&#8203;:17AM \-0700\, david @&#8203; profane\. mongueurs\. net wrote&#8203;:
> > 
> > > panic&#8203;: fold\_constants JMPENV\_PUSH returned 2\.
> > 
> > Mmm\, yes\. I think the minimal test case is&#8203;:
> > 
> > $ \./perl \-lwe 'BEGIN \{$SIG\{\_\_WARN\_\_\} = sub\{exit\}\} "a" == "b"'
> > panic&#8203;: fold\_constants JMPENV\_PUSH returned 2\.
> > 
> > Presumably the constant folding code should ignore exit\, as it's only an
> > optimisation? ie this could should print "Here\!"&#8203;:
> > 
> > perl \-lwe 'BEGIN \{$SIG\{\_\_WARN\_\_\} = sub\{exit\}\} if \(0\) \{"a" == "b"\}; print "Here\!"'
> 
> We've talked before about bailing out of constant\-folding on a croak;
> maybe do the same for any warning?  \(Harder to do\, though\.\)

now fixed byt change \#28148

\-\- 
A major Starfleet emergency breaks out near the Enterprise\, but
fortunately some other ships in the area are able to deal with it to
everyone's satisfaction\.
&nbsp;    \-\- Things That Never Happen in "Star Trek" \#13

**Comment by p5pRT** [comment]
@iabyn - Status changed from 'open' to 'resolved'
